lib/ukboot/early_init.c

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/argparse.h>
+#include <uk/assert.h>
+#include <uk/boot/earlytab.h>
+#include <uk/essentials.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/prio.h>
+
+#ifdef CONFIG_LIBUKLIBPARAM
+#include <uk/libparam.h>
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+char *argv[CONFIG_LIBUKBOOT_MAXNBARGS];
+int argc;
+
+static char cmdline[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
+static __sz cmdline_len;
+
+#ifdef CONFIG_LIBUKLIBPARAM
+static int init_libparam(int argc, char *argv[])
+{
+	int rc;
+	/*
+	 * First, we scan if we can find the stop sequence in the kernel
+	 * cmdline. If not, we assume that there are no uklibparam arguments in
+	 * the command line.
+	 * NOTE: argv[0] contains the kernel/program name that we need to
+	 *       hide from the parser.
+	 */
+	rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_SCAN);
+	if (rc > 0 && rc < (argc - 1)) {
+		/* In this case, we did successfully scan for uklibparam
+		 * arguments and stop sequence is at rc < (argc - 1).
+		 */
+		/* Run a second pass for parsing */
+		rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_USAGE);
+		if (rc < 0) /* go down on errors (including USAGE) */
+			return -EINVAL;
+
+		/* Drop uklibparam parameters from argv but keep argv[0].
+		 * We are going to replace the stop sequence with argv[0].
+		 */
+		rc += 1; /* include argv[0]; we use rc as idx to stop seq */
+		argc -= rc;
+		argv[rc] = argv[0];
+		argv = &argv[rc];
+	}
+	return 0;
+}
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+static int uk_boot_early_init_parse_cmdl(struct ukplat_bootinfo *bi)
+{
+	const char *cmdl;
+	int rc __maybe_unused;
+
+	UK_ASSERT(bi);
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
+		return -ENOTSUP;
+
+	/* Copy the commandline into a mutable buffer.
+	 * We will pass this to ukargparse to thrash.
+	 */
+	if (!bi->cmdline_len) {
+		cmdl = CONFIG_UK_NAME;
+		cmdline_len = sizeof(CONFIG_UK_NAME);
+	} else {
+		cmdl = (const char *)bi->cmdline;
+		cmdline_len = bi->cmdline_len;
+	}
+
+	if (cmdline_len > ARRAY_SIZE(cmdline))
```

Comment:
`unlikely`? And why `ARRAY_SIZE` instead of `sizeof`?

Reply:
`unlikely` ok. Why *not* `ARRAY_SIZE`?

Reply:
I like `sizeof` better because it is built-in and makes it clear we are dealing with a string. Maybe it is just personal preference though?

Reply:
Yeah, I think so as others may argue that it's safer to always use `ARRAY_SIZE` regardless of the underlying type.

Reply:
Ok fine. I just find it unusual being used for strings.